pkg/process/procutil/process_linux.go

```diff
@@ -0,0 +1,173 @@
+// +build linux
+
+package procutil
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/DataDog/datadog-agent/pkg/process/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/gopsutil/host"
+)
+
+// Probe is a service that fetches process related info on current host
+type Probe struct {
+	procRootLoc  string // ProcFS
+	procRootFile *os.File
+
+	uid  uint32 // Used for path permission checking to prevent access to files that we can't access
+	euid uint32
+
+	bootTime uint64
+}
+
+// NewProcessProbe initializes a new Probe object
+func NewProcessProbe() *Probe {
+	bootTime, _ := host.BootTime() // TODO (sk): Rewrite this w/o gopsutil
+
+	p := &Probe{
+		procRootLoc: util.HostProc(),
+		uid:         uint32(os.Getuid()),
+		euid:        uint32(os.Geteuid()),
+		bootTime:    bootTime,
+	}
+	return p
+}
+
+// Close cleans up everything related to Probe object
+func (p *Probe) Close() {
+	if p.procRootFile != nil {
+		p.procRootFile.Close()
+		p.procRootFile = nil
+	}
+}
+
+// ProcessesByPID returns a map of process info indexed by PID
+func (p *Probe) ProcessesByPID() (map[int32]*Process, error) {
+	pids, err := p.getActivePIDs()
+	if err != nil {
+		return nil, err
+	}
+
+	procsByPID := make(map[int32]*Process, len(pids))
+	for _, pid := range pids {
+		pathForPID := filepath.Join(p.procRootLoc, strconv.Itoa(int(pid)))
+		if !util.PathExists(pathForPID) {
+			log.Debugf("Unable to create new process %d, dir doesn't exist", pid)
+			continue
+		}
+
+		cmdline := p.getCmdline(pathForPID)
+		if len(cmdline) == 0 {
```

Comment:
We're returning `nil` in `getCmdline` if `len(cmdline, in bytes) == 0`. Executing `len(nil)` here won't cause a problem?